design/crypto_primitives.rs

```diff
@@ -0,0 +1,179 @@
+            from()
+        }
+        /// Occurs when we fail to derive encryption key from password.
+        DeriveKey {
+            description("error deriving encryption key from password")
+        }
+    }
+}
+
+/// Derives encryption key from plaintext password.
+pub fn key_from_password(password: &[u8], salt: &[u8]) -> Result<SymmetricKey, Error>;
+
+// no signing key
+// Make it `Copy`able?
+#[derive(Debug, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize, Clone)]
+pub struct PublicEncryptKey {
+    encrypt: box_::PublicKey,
+}
+
+impl PublicEncryptKey {
+    // No new() method, since public key should be constructed using `SecretEncryptKey`
+    // pub fn new() -> Self
+
+    /// Construct public key from bytes. Useful when it was serialized before.
+    pub fn from_bytes(public_key: [u8; PUBLIC_ENCRYPT_KEY_BYTES]) -> Result<Self, Error>
+
+    /// For anyone who wants to store public key.
+    pub fn as_bytes(self) -> [u8; PUBLIC_ENCRYPT_KEY_BYTES]
+
+    pub fn encrypt_anonymous<T: Serialize>(&self, plaintext: &T) -> Result<Vec<u8>, Error>
+
+    pub fn encrypt_anonymous_bytes(&self, plaintext: &[u8]) -> Vec<u8> {
+}
+
+/// Store secret key in `Arc` to discourage cloning it. Makes it more secure by not duplicating
+/// encryption keys in memory.
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct SecretEncryptKey {
+    inner: Arc<SecretEncryptKeyInner>,
+    public: PublicKey,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+struct SecretEncryptKeyInner {
+    encrypt: box_::SecretKey,
+}
+
+impl SecretEncryptKey {
+    /// Construct random public/private encryption key pair.
+    pub fn new() -> Self
+
+    /// Construct secret key from given bytes. Useful when secret key was serialized before.
+    // Note that we don't need to provide public key, seems like it's derived from secret key:
+    // https://docs.rs/rust_sodium/0.9.0/rust_sodium/crypto/box_/curve25519xsalsa20poly130/struct.SecretKey.html#method.public_key
+    pub fn from_bytes(secret_key: [u8; SECRET_ENCRYPT_KEY_BYTES]) -> Result<Self, Error>
+
+    /// Returns the public part of the secret key set.
+    pub fn pub_key(&self) -> &PublicEncryptKey
+
+    /// Derive shared key for symmetric data encryption
+    pub fn shared_secret(&self, their_pk: &PublicEncryptKey) -> SharedSecretKey
+
+    /// For anyone who wants to store secret key.
+    pub fn as_bytes(self) -> [u8; SECRET_ENCRYPT_KEY_BYTES]
+
+    pub fn decrypt_anonymous<T: Serialize + DeserializeOwned>(&self, ciphertext: &[u8]) -> Result<T, Error>
+
+    pub fn decrypt_anonymous_bytes(&self, ciphertext: &[u8]) -> Result<Vec<u8>, Error>
+}
+
+#[derive(Debug, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize, Clone)]
+pub struct PublicSignKey {
+    sign: sign::PublicKey,
+}
+
+impl PublicSignKey {
+    pub fn verify_detached(&self, signature: &Signature, data: &[u8]) -> bool {
+
+    pub fn from_bytes(public_key: [u8; PUBLIC_SIGN_KEY_BYTES]) -> Result<Self, Error>
```

Comment:
1. In context of asymmetric keys, when would we ever error out here ? Given a secret we could know the public key in some standards i think, not sure that if given a public key can you actually infer anything, including if it's a part of valid keypair ? Is this the zero-knowledge proof thingy wherein I can prove that this is the product of two big primes without revealing the primes, or something ?
1. if the above point with zero knowledge proof etc. is valid, is this extensible to all algos (like RSA/ECDH etc) ?

My knowledge is sparse here - might need confirming from others if the API is fine.
Similarly in other asymmetric keys' `from_bytes()` functions here.

Reply:
You're right, I don't see ways to validate bytes as a key. I've looked into a couple of ciphers (chacha20 and salsa20) from rust-crypto crate and it doesn't do any validation, just constructs structure from given bytes.